File: plugins/expo-autofill-plugin/ios-template/PearPassAutofillExtension/PasskeyJobCreator.swift
//
//  PasskeyJobCreator.swift
//  PearPassAutoFillExtension
//
//  Creates job queue entries for passkey registration and updates.
//  Instead of writing directly to the vault (which requires exclusive
//  file locks held by the main app), jobs are encrypted and saved to
//  a shared file. The main app processes them on next launch/resume.
//

import Foundation

// MARK: - Errors

enum PasskeyJobError: LocalizedError {
    case noHashedPassword
    case invalidHashedPassword
    case jobCreationFailed(String)

    var errorDescription: String? {
        switch self {
        case .noHashedPassword:
            return "No hashed password available from vault"
        case .invalidHashedPassword:
            return "Hashed password has invalid format or length"
        case .jobCreationFailed(let reason):
            return "Failed to create job: \(reason)"
        }
    }
}

// MARK: - PasskeyJobCreator

enum PasskeyJobCreator {

    // MARK: - Create ADD_PASSKEY Job

    /// Creates an ADD_PASSKEY job, saves any file attachments, and appends to the job queue.
    ///
    /// - Parameters:
    ///   - vaultId: The target vault ID.
    ///   - credential: The generated passkey credential.
    ///   - formData: The user-edited form data (title, websites, note, folder, attachments).
    ///   - rpId: Relying party identifier (e.g. "github.com").
    ///   - rpName: Relying party display name.
    ///   - userId: WebAuthn user handle (base64URL).
    ///   - userName: WebAuthn user name.
    ///   - userDisplayName: WebAuthn user display name.
    ///   - hashedPassword: The 32-byte encryption key from vault masterEncryption.
    /// - Returns: The UUID of the created job.
    /// - Throws: `PasskeyJobError`, `JobFileError`, or `JobEncryptionError`.
    static func createAddPasskeyJob(
        vaultId: String,
        credential: PasskeyCredential,
        formData: PasskeyFormData,
        rpId: String,
        rpName: String,
        userId: String,
        userName: String,
        userDisplayName: String,
        hashedPassword: Data
    ) throws -> String {
        let jobId = UUID().uuidString
        let now = Date().timeIntervalSince1970 * 1000 // milliseconds
        let recordId = formData.existingRecord?.id ?? UUID().uuidString

        // Save file attachments to pearpass_jobs/attachments/
        var jobAttachments: [JobAttachment] = []
        for attachment in formData.attachments {
            let relativePath = try JobFileManager.saveAttachment(
                data: attachment.data,
                attachmentId: attachment.id,
                originalFilename: attachment.name
            )
            jobAttachments.append(JobAttachment(
                id: attachment.id,
                name: attachment.name,
                relativePath: relativePath
            ))
        }

        // Normalize websites
        var websites: [String] = formData.websites
        if websites.isEmpty {
            websites = ["https://\(rpId)"]
        }

        // Build payload
        let payload = AddPasskeyPayload(
            rpId: rpId,
            rpName: rpName,
            userId: userId,
            userName: userName,
            userDisplayName: userDisplayName,
            credentialId: credential.id,
            publicKey: credential.response.publicKey,
            privateKey: credential._privateKeyBuffer.base64URLEncodedString(),
            clientDataJSON: credential.response.clientDataJSON,
            attestationObject: credential.response.attestationObject,
            authenticatorData: credential.response.authenticatorData,
            algorithm: credential.response.publicKeyAlgorithm,
            createdAt: Double(formData.passkeyCreatedAt),
            transports: credential.response.transports,
            recordId: recordId,
            title: formData.title,
            note: formData.note.isEmpty ? nil : formData.note,
            folder: formData.folder,
            websites: websites,
            attachments: jobAttachments
        )

        // Build job
        let job = Job(
            id: jobId,
            type: .addPasskey,
            status: .pending,
            createdAt: now,
            updatedAt: now,
            retryCount: 0,
            maxRetries: 3,
            vaultId: vaultId,
            payload: .addPasskey(payload),
            error: nil
        )

        // Append to encrypted queue
        try JobFileManager.appendJob(job, hashedPassword: hashedPassword)

        NSLog("[PasskeyJobCreator] Created ADD_PASSKEY job \(jobId) for vault \(vaultId)")
        return jobId
    }

    // MARK: - Create UPDATE_PASSKEY Job

    /// Creates an UPDATE_PASSKEY job and appends it to the job queue.
    /// Used when adding a passkey to an existing login record.
    ///
    /// - Parameters:
    ///   - vaultId: The target vault ID.
    ///   - existingRecordId: The ID of the existing record to merge the passkey into.
    ///   - credential: The generated passkey credential.
    ///   - rpId: Relying party identifier.
    ///   - rpName: Relying party display name.
    ///   - userId: WebAuthn user handle (base64URL).
    ///   - userName: WebAuthn user name.
    ///   - userDisplayName: WebAuthn user display name.
    ///   - hashedPassword: The 32-byte encryption key from vault masterEncryption.
    ///   - passkeyCreatedAt: Timestamp (ms) when the passkey was created.
    /// - Returns: The UUID of the created job.
    /// - Throws: `PasskeyJobError`, `JobFileError`, or `JobEncryptionError`.
    static func createUpdatePasskeyJob(
        vaultId: String,
        existingRecordId: String,
        credential: PasskeyCredential,
        rpId: String,
        rpName: String,
        userId: String,
        userName: String,
        userDisplayName: String,
        hashedPassword: Data,
        passkeyCreatedAt: Int64
    ) throws -> String {
        let jobId = UUID().uuidString
        let now = Date().timeIntervalSince1970 * 1000

        let payload = UpdatePasskeyPayload(
            existingRecordId: existingRecordId,
            rpId: rpId,
            rpName: rpName,
            userId: userId,
            userName: userName,
            userDisplayName: userDisplayName,
            credentialId: credential.id,
            publicKey: credential.response.publicKey,
            privateKey: credential._privateKeyBuffer.base64URLEncodedString(),
            clientDataJSON: credential.response.clientDataJSON,
            attestationObject: credential.response.attestationObject,
            authenticatorData: credential.response.authenticatorData,
            algorithm: credential.response.publicKeyAlgorithm,
            createdAt: Double(passkeyCreatedAt),
            transports: credential.response.transports,
            vaultId: vaultId
        )

        let job = Job(
            id: jobId,
            type: .updatePasskey,
            status: .pending,
            createdAt: now,
            updatedAt: now,
            retryCount: 0,
            maxRetries: 3,
            vaultId: vaultId,
            payload: .updatePasskey(payload),
            error: nil
        )

        try JobFileManager.appendJob(job, hashedPassword: hashedPassword)

        NSLog("[PasskeyJobCreator] Created UPDATE_PASSKEY job \(jobId) for record \(existingRecordId)")
        return jobId
    }

    // MARK: - Get Hashed Password

    /// Fetches the hashed password from the vault's masterEncryption metadata.
    /// The hashed password is a hex-encoded string representing a 32-byte key.
    ///
    /// - Parameter vaultClient: An initialized and authenticated vault client.
    /// - Returns: The 32-byte hashed password as `Data`, or `nil` if unavailable.
    static func getHashedPassword(from vaultClient: PearPassVaultClient) async -> Data? {
        do {
            let masterEncryptionData = try await vaultClient.vaultsGet(key: "masterEncryption")

            guard let hashedPasswordHex = masterEncryptionData["hashedPassword"] as? String else {
                NSLog("[PasskeyJobCreator] No hashedPassword in masterEncryption")
                return nil
            }

            // The hashedPassword is stored as a hex string — convert to raw bytes
            guard let data = Data(hexString: hashedPasswordHex) else {
                NSLog("[PasskeyJobCreator] Failed to decode hashedPassword hex string")
                return nil
            }

            guard data.count == JobEncryption.keyBytes else {
                NSLog("[PasskeyJobCreator] hashedPassword is \(data.count) bytes, expected \(JobEncryption.keyBytes)")
                return nil
            }

            return data
        } catch {
            NSLog("[PasskeyJobCreator] Error fetching masterEncryption: \(error.localizedDescription)")
            return nil
        }
    }
}

// MARK: - Hex String Extension

extension Data {
    /// Initialize Data from a hex-encoded string.
    /// Returns nil if the string contains non-hex characters or has odd length.
    init?(hexString: String) {
        let hex = hexString.lowercased()
        guard hex.count % 2 == 0 else { return nil }

        var data = Data(capacity: hex.count / 2)
        var index = hex.startIndex

        while index < hex.endIndex {
            let nextIndex = hex.index(index, offsetBy: 2)
            let byteString = hex[index..<nextIndex]
            guard let byte = UInt8(byteString, radix: 16) else { return nil }
            data.append(byte)
            index = nextIndex
        }

        self = data
    }

    /// Returns a hex-encoded string representation of the data.
    func hexEncodedString() -> String {
        return map { String(format: "%02x", $0) }.joined()
    }
}

File: plugins/expo-autofill-plugin/android-template/java/autofill/jobs/PasskeyJobCreator.java
package com.pears.pass.autofill.jobs;

import com.pears.pass.autofill.crypto.Base64URLUtils;
import com.pears.pass.autofill.data.PasskeyCredential;
import com.pears.pass.autofill.data.PasskeyFormData;
import com.pears.pass.autofill.utils.SecureLog;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

/**
 * Creates ADD_PASSKEY and UPDATE_PASSKEY jobs for the job queue.
 *
 * This is the primary entry point for the autofill service to enqueue passkey
 * write operations. Instead of writing directly to the vault (which would fail
 * due to file locking), the autofill service creates a job that the main app
 * will process on resume.
 *
 * Usage from PasskeyRegistrationActivity:
 * <pre>
 *   PasskeyJobCreator creator = new PasskeyJobCreator(new JobFileManager(context));
 *   String jobId = creator.createAddPasskeyJob(
 *       vaultId, credential, formData,
 *       rpId, rpName, userId, userName, userDisplayName,
 *       hashedPassword
 *   );
 * </pre>
 */
public class PasskeyJobCreator {
    private static final String TAG = "PasskeyJobCreator";

    private final JobFileManager jobFileManager;

    public PasskeyJobCreator(JobFileManager jobFileManager) {
        this.jobFileManager = jobFileManager;
    }

    /**
     * Create an ADD_PASSKEY job: saves attachments, builds the payload, and appends
     * an encrypted job to the queue file.
     *
     * @param vaultId          Target vault ID
     * @param credential       The generated PasskeyCredential (contains credentialId, publicKey, privateKey, userId, response)
     * @param formData         User-entered form data (title, username, websites, note, folder, attachments)
     * @param rpId             Relying party ID (e.g., "github.com")
     * @param rpName           Relying party display name
     * @param userId           User handle as base64URL
     * @param userName         User name
     * @param userDisplayName  User display name
     * @param hashedPassword   32-byte encryption key from vault masterEncryption
     * @return The job ID (UUID string)
     * @throws Exception if attachment saving, encryption, or file I/O fails
     */
    public String createAddPasskeyJob(
            String vaultId,
            PasskeyCredential credential,
            PasskeyFormData formData,
            String rpId, String rpName,
            String userId, String userName, String userDisplayName,
            byte[] hashedPassword) throws Exception {

        SecureLog.d(TAG, "Creating ADD_PASSKEY job for rpId: " + rpId);

        // 1. Save attachment files to pearpass_jobs/attachments/
        List<JobAttachment> attachments = new ArrayList<>();
        if (formData.getAttachments() != null) {
            for (PasskeyFormData.AttachmentFile attachment : formData.getAttachments()) {
                String attachmentId = UUID.randomUUID().toString();
                String relativePath = jobFileManager.saveAttachment(
                        attachment.getData(), attachmentId, attachment.getName());
                attachments.add(new JobAttachment(attachmentId, attachment.getName(), relativePath));
            }
        }

        // 2. Build the payload
        String recordId = UUID.randomUUID().toString();
        long createdAt = System.currentTimeMillis();

        AddPasskeyPayload payload = new AddPasskeyPayload(
                rpId, rpName,
                userId, userName, userDisplayName,
                credential.getId(),
                credential.getResponse().getPublicKey(),
                Base64URLUtils.encode(credential.getPrivateKeyBuffer()),
                credential.getResponse().getClientDataJSON(),
                credential.getResponse().getAttestationObject(),
                credential.getResponse().getAuthenticatorData(),
                credential.getResponse().getPublicKeyAlgorithm(),
                createdAt,
                credential.getResponse().getTransports(),
                recordId,
                formData.getTitle(),
                formData.getNote(),
                formData.getFolder(),
                formData.getWebsites(),
                attachments
        );

        // 3. Create the job and append to the encrypted queue
        Job job = new Job(
                UUID.randomUUID().toString(),
                Job.JobType.ADD_PASSKEY,
                Job.JobStatus.PENDING,
                createdAt,
                0,   // retryCount
                3,   // maxRetries
                vaultId,
                payload.toJSON()
        );

        jobFileManager.appendJob(job, hashedPassword);

        SecureLog.d(TAG, "Created ADD_PASSKEY job " + job.getId()
                + " with " + attachments.size() + " attachment(s)");
        return job.getId();
    }

    /**
     * Create an UPDATE_PASSKEY job: builds the payload and appends an encrypted job
     * to the queue file. No attachments are saved for update jobs.
     *
     * @param vaultId          Target vault ID
     * @param existingRecordId ID of the existing login record to merge the passkey into
     * @param credential       The generated PasskeyCredential
     * @param rpId             Relying party ID
     * @param rpName           Relying party display name
     * @param userId           User handle as base64URL
     * @param userName         User name
     * @param userDisplayName  User display name
     * @param hashedPassword   32-byte encryption key from vault masterEncryption
     * @return The job ID (UUID string)
     * @throws Exception if encryption or file I/O fails
     */
    public String createUpdatePasskeyJob(
            String vaultId,
            String existingRecordId,
            PasskeyCredential credential,
            String rpId, String rpName,
            String userId, String userName, String userDisplayName,
            byte[] hashedPassword) throws Exception {

        SecureLog.d(TAG, "Creating UPDATE_PASSKEY job for record: " + existingRecordId);

        long createdAt = System.currentTimeMillis();

        // Build the payload
        UpdatePasskeyPayload payload = new UpdatePasskeyPayload(
                existingRecordId,
                rpId, rpName,
                userId, userName, userDisplayName,
                credential.getId(),
                credential.getResponse().getPublicKey(),
                Base64URLUtils.encode(credential.getPrivateKeyBuffer()),
                credential.getResponse().getClientDataJSON(),
                credential.getResponse().getAttestationObject(),
                credential.getResponse().getAuthenticatorData(),
                credential.getResponse().getPublicKeyAlgorithm(),
                createdAt,
                credential.getResponse().getTransports(),
                vaultId
        );

        // Create the job and append to the encrypted queue
        Job job = new Job(
                UUID.randomUUID().toString(),
                Job.JobType.UPDATE_PASSKEY,
                Job.JobStatus.PENDING,
                createdAt,
                0,   // retryCount
                3,   // maxRetries
                vaultId,
                payload.toJSON()
        );

        jobFileManager.appendJob(job, hashedPassword);

        SecureLog.d(TAG, "Created UPDATE_PASSKEY job " + job.getId()
                + " for existing record " + existingRecordId);
        return job.getId();
    }
}
